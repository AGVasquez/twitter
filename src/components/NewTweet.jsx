import React, { Component, useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTweet as addTweetAction } from '../actions';
import Avatar from './Avatar';
import { isEmpty } from '../utils';

// const MAX_CHARS = 60; // TODO: Implement max for input

function useLocalStorageState(
  key,
) {
  const valueInLocalStorage = window.localStorage.getItem(key)

  const[text,setText]= useState( valueInLocalStorage || '')

  const persistThings = useCallback( (key,value) =>{
    window.localStorage.setItem(key,value)
  },[text])

  useEffect(()=>{
    persistThings(key,text)
  },[text])

  return [text,setText]
}

const NewTweet = ({ user, addTweet }) => {

  const [text,setText] = useLocalStorageState('persitedText')
  const publishTweet= useCallback(()=> {
    addTweet({
      content: text,
      userId: user?.id,
      likes: [],
      date: new Date(),
      retweets: 0,
    });
  },[text,user]
  )

    if (!user || isEmpty(user)) {
      return null;
    }

    return (
      <div className="new-tweet">
        <Avatar src={user.avatar} />
        <input
          value={text}
          className="new-tweet-input"
          placeholder="What's happening?"
          data-testid="new-tweet-input"
          onChange={({ target: { value } }) => setText( value )}
        />
        <button
          className="new-tweet-button"
          type="button"
          data-testid="new-tweet-button"
          onClick={publishTweet}
        >
          Tweet
        </button>
      </div>
    );
  
}

NewTweet.propTypes = {
  user: PropTypes.object.isRequired,
  addTweet: PropTypes.func.isRequired,
};

const mapStateToProps = ({ user }) => ({
  user,
});

export default connect(mapStateToProps, { addTweet: addTweetAction })(NewTweet);
