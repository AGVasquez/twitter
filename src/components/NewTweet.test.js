import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { NewTweet } from './NewTweet';

describe('New tweet', () => {
  test('it renders correctly', () => {
    const { container } = render(<NewTweet />);
    expect(container).toMatchSnapshot();
  });
  test.skip('Enforce max lenght for tweet', () => {
    const { container, getByTestId } = render(<NewTweet />);
    const input = getByTestId('new-tweet-input');
    const button = getByTestId('new-tweet-button');
    // TODO: Try to type more than max number of chars and validate the result
  });
});
