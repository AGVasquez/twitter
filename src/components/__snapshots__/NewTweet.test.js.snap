// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New tweet it renders correctly 1`] = `<div />`;
